types/clevertap-web-sdk/clevertap.d.ts

```diff
@@ -0,0 +1,111 @@
+type Region = 'eu1' | 'in1' | 'sg1' | 'us1' | 'sk1';
+
+interface PrivacyData {
+    optOut?: boolean;
+    useIP?: boolean;
+}
+
+interface Privacy extends Array<any> {
+    push(...privacyArr: PrivacyData[]): 0;
+}
+
+type EventName = string;
+type EventData = object;
+type EventNameOrData = EventName | EventData;
+interface EventDetails {
+    firstTime: Date;
+    lastTime: Date;
+    count: number;
+}
+interface EventHandler extends Array<any> {
+    push(evtName: string, ...evtNameOrData: EventNameOrData[]): 0;
+    getDetails(evtName: string): EventDetails | undefined;
+}
+
+interface SiteData {
+    Name?: string;
+    Identity?: string | number;
+    Gender?: 'M' | 'F';
+    Employed?: 'Y' | 'N';
+    Married?: 'Y' | 'N';
+    Education?: 'School' | 'College' | 'Graduate';
+    Age?: string | number;
+    DOB?: string | number | Date;
+    Phone?: string | number;
+    [key: string]: any;
+}
+interface ProfileData {
+    Site?: SiteData;
+    Facebook?: object;
+    'Google Plus'?: object;
+}
+interface ProfileHandler extends Array<any> {
+    push(...profileData: ProfileData[]): 0;
+    getAttribute(profileName: string): any;
+}
+
+interface UserLoginHandler extends Array<any> {
+    push(...profileData: ProfileData[]): 0;
+    clear(): void;
+}
+
+interface NotificationData {
+    titleText: string;
+    bodyText: string;
+    okButtonText: string;
+    rejectButtonText: string;
+    okButtonColor?: string;
+    skipDialog?: boolean;
+    askAgainTimeInSeconds?: number;
+    okCallback?: () => void;
+    rejectCallback?: () => void;
+    subscriptionCallback?: () => void;
+    hidePoweredByCT?: boolean;
+    serviceWorkerPath?: string;
+    httpsPopupPath?: string;
+    httpsIframePath?: string;
+    apnsWebPushId?: string;
+    apnsWebPushServiceUrl?: string;
+}
+
+interface NotificationHandler extends Array<any> {
+    push(notificationData: NotificationData): 0;
+    push(
+        titleText: string,
+        bodyText: string,
+        okButtonText: string,
+        rejectButtonText: string,
+        okButtonColor?: string,
+        skipDialog?: boolean,
+        askAgainTimeInSeconds?: number,
+    ): 0;
+}
+
+interface User {
+    getTotalVisits(): number | undefined;
+    getLastVisit(): Date | undefined;
+}
+
+interface Session {
+    getTimeElapsed(): number | undefined;
+    getPageCount(): number | undefined;
+}
+
+declare const clevertap: {
```

Comment:
this looks wrong, this is a class symbol:
https://github.com/CleverTap/clevertap-web-sdk/blob/master/src/clevertap.js#L29
it should support `instanceof` narrowing and other language features

Reply:
Created an interface of `CleverTap` and used it as the type of `clevertap`. The `CleverTap` class isn't exported from main.js, only an instance is exported. https://github.com/CleverTap/clevertap-web-sdk/blob/master/src/main.js

Reply:
Im only pointing from outside onlooker, the code here should look like:
```ts
export default class CleverTap {
```
class could have no constructor defined, as it's not exported later
so the `index.d.ts` (being UMD), could  look like real code in JS module:
```ts
import CleverType from './clevertap';

declare const clevertap: CleverType;
export = clevertap;

export as namespace clevertap;
```